katsdpcontroller/test/test_product_config.py

```diff
@@ -1,73 +1,1012 @@
+    )
+
+
+class TestContinuumImageStream:
+    """Test :class:`~.ContinuumImageStream`."""
+
+    def setup(self) -> None:
+        self.flags = make_flags()
+        self.config: Dict[str, Any] = {
+            'type': 'sdp.continuum_image',
+            'src_streams': ['sdp_l0_continuum_flags'],
+            'uvblavg_parameters': {'foo': 'bar'},
+            'mfimage_parameters': {'abc': 123},
+            'max_realtime': 10000.0,
+            'min_time': 100.0
+        }
+
+    def test_from_config(self) -> None:
+        cont = ContinuumImageStream.from_config(
+            Options(), 'continuum_image', self.config, [self.flags], {}
+        )
+        assert_equal(cont.name, 'continuum_image')
+        # Make sure that the value is copied
+        self.config['uvblavg_parameters'].clear()
+        self.config['mfimage_parameters'].clear()
+        assert_equal(cont.uvblavg_parameters, {'foo': 'bar'})
+        assert_equal(cont.mfimage_parameters, {'abc': 123})
+        assert_equal(cont.max_realtime, 10000.0)
+        assert_equal(cont.min_time, 100.0)
+        assert_is(cont.flags, self.flags)
+        assert_is(cont.cal, self.flags.cal)
+        assert_is(cont.vis, self.flags.vis)
+
+    def test_defaults(self) -> None:
+        del self.config['uvblavg_parameters']
+        del self.config['mfimage_parameters']
+        del self.config['max_realtime']
+        del self.config['min_time']
+        cont = ContinuumImageStream.from_config(
+            Options(), 'continuum_image', self.config, [self.flags], {}
+        )
+        assert_equal(cont.uvblavg_parameters, {})
+        assert_equal(cont.mfimage_parameters, {})
+        assert_equal(cont.max_realtime, None)
+        assert_equal(cont.min_time, defaults.CONTINUUM_MIN_TIME)
+
+
+def make_continuum_image(flags: Optional[FlagsStream] = None) -> ContinuumImageStream:
+    if flags is None:
+        flags = make_flags()
+    return ContinuumImageStream(
+        'continuum_image', [flags],
+        uvblavg_parameters={},
+        mfimage_parameters={},
+        max_realtime=10000.0,
+        min_time=100.0
+    )
+
+
+class TestSpectralImageStream:
+    """Test :class:`~.SpectralImageStream`."""
+
+    def setup(self) -> None:
+        self.flags = make_flags()
+        self.continuum_image = make_continuum_image()
+        self.config: Dict[str, Any] = {
+            'type': 'sdp.spectral_image',
+            'src_streams': ['sdp_l0_continuum_flags', 'continuum_image'],
+            'output_channels': [64, 100],
+            'min_time': 200.0,
+            'parameters': {'q_fov': 2.0}
+        }
+
+    def test_from_config(self) -> None:
+        spec = SpectralImageStream.from_config(
+            Options(), 'spectral_image', self.config, [self.flags, self.continuum_image], {}
+        )
+        # Check that it gets copied
+        self.config['parameters'].clear()
+        assert_equal(spec.name, 'spectral_image')
+        assert_equal(spec.output_channels, (64, 100))
```

Comment:
Wouldn't it be more robust to compare to the config dict - assert_equal(spec.output_channels, self.config['output_channels']) ?

Reply:
Done